internal/client/consensus/grandpa/communication_gossip.go

```diff
@@ -0,0 +1,1708 @@
+	}
+
+	// only r-1 ... r+1
+	if round > saturating.Add(v.round, 1) {
+		return considerRejectFuture
+	} else if round < saturating.Sub(v.round, 1) {
+		return considerRejectPast
+	}
+
+	return considerAccept
+}
+
+// Consider a set-id global message. Rounds are not taken into account, but are implicitly because we gate on
+// finalization of a further block than a previous commit.
+func (v view[N]) considerGlobal(setID SetID, number N) consider {
+	// only from current set
+	if setID < v.setID {
+		return considerRejectPast
+	} else if setID > v.setID {
+		return considerRejectFuture
+	}
+
+	// only commits which claim to prove a higher block number than the one we're aware of.
+	if v.lastCommit == nil {
+		return considerAccept
+	}
+	if *v.lastCommit < number {
+		return considerAccept
+	}
+	return considerRejectPast
+}
+
+type numberRoundSetID[N runtime.Number] struct {
+	Number N
+	Round
+	SetID
+}
+
+// A local view of protocol state. Similar to `View` but we additionally track the round and set id at which the last
+// commit was observed, and the instant at which the current round started.
+type localView[N runtime.Number] struct {
+	round      Round
+	setID      SetID
+	lastCommit *numberRoundSetID[N]
+	roundStart time.Time
+}
+
+func newLocalView[N runtime.Number](setID SetID, round Round) *localView[N] {
+	return &localView[N]{
+		round:      round,
+		setID:      setID,
+		lastCommit: nil,
+		roundStart: time.Now(),
+	}
+}
+
+// Converts the local view to a `view` discarding round and set id information about the last commit.
+func (lv *localView[N]) view() view[N] {
+	return view[N]{
+		round:      lv.round,
+		setID:      lv.setID,
+		lastCommit: lv.lastCommitHeight(),
+		lastUpdate: nil,
+	}
+}
+
+// Update the set ID. implies a reset to round 1.
+func (lv *localView[N]) updateSet(setID SetID) {
+	if setID != lv.setID {
+		lv.setID = setID
+		lv.round = 1
+		lv.roundStart = time.Now()
+	}
+}
+
+// Updates the current round.
+func (lv *localView[N]) updateRound(round Round) {
+	lv.round = round
+	lv.roundStart = time.Now()
+}
+
+// Returns the height of the block that the last observed commit finalizes.
+func (lv *localView[N]) lastCommitHeight() *N {
+	if lv.lastCommit != nil {
+		return &lv.lastCommit.Number
+	}
+	return nil
+}
+
+const keepRecentRounds uint = 3
+
+type roundSetID struct {
+	Round
+	SetID
+}
+
+type keepTopicsMapEntry struct {
+	*Round
+	SetID
+}
+
+// Tracks gossip topics that we are keeping messages for. We keep topics of:
+//   - the last `keepRecentRounds` complete GRANDPA rounds,
+//   - the topic for the current and next round,
+//   - and a global topic for commit and catch-up messages.
+type keepTopics[H runtime.Hash, Hasher runtime.Hasher[H]] struct {
+	currentSet SetID
+	rounds     deque.Deque[roundSetID]
+	reverseMap map[H]keepTopicsMapEntry
+}
+
+func newKeepTopics[H runtime.Hash, Hasher runtime.Hasher[H]]() keepTopics[H, Hasher] {
+	var dq deque.Deque[roundSetID]
+	dq.SetBaseCap(3 + 2)
```

Comment:
why not just 5 and can we make it a constant with a meaningful name?
```suggestion
	dq.SetBaseCap(5)
```

Reply:
Whoops! Supposed to be `keepRecentRounds + 2`.